sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1818,7 +1818,10 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             "sqlExpr" -> a.sql,
             "cols" -> cols))
       }
-      resolved
+      resolved match {
```

Comment:
This change is needed as references to nested columns in MERGE are resolved like `a.nested_col.i AS i` (with added aliases), which breaks our assumption about what kind of keys we expect. Let me know if anyone can spot any edge cases when it is not safe.